src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -33,20 +33,24 @@ public partial class Solution
         /// <summary>
         /// Result of calling <see cref="WithFrozenPartialCompilationsAsync"/>.
         /// </summary>
-        private AsyncLazy<Solution> CachedFrozenSolution { get; init; }
+        private readonly AsyncLazy<Solution> _cachedFrozenSolution;
 
         /// <summary>
         /// Mapping of DocumentId to the frozen solution we produced for it the last time we were queried.  This
         /// instance should be used as its own lock when reading or writing to it.
         /// </summary>
         private readonly Dictionary<DocumentId, AsyncLazy<Solution>> _documentIdToFrozenSolution = [];
 
-        private Solution(SolutionCompilationState compilationState)
+        private Solution(
+            SolutionCompilationState compilationState,
+            AsyncLazy<Solution>? cachedFrozenSolution = null)
         {
             _projectIdToProjectMap = [];
             _compilationState = compilationState;
 
-            this.CachedFrozenSolution = AsyncLazy.Create(ComputeFrozenSolution);
+            _cachedFrozenSolution = cachedFrozenSolution ?? new AsyncLazy<Solution>(
+                c => Task.FromResult(ComputeFrozenSolution(c)),
+                c => ComputeFrozenSolution(c));
```

Comment:
Can this use your other helper? If this is coming in a queued PR then feel free to resolve this comment.

Reply:
will do!